### ISSUE TITLE
FSRS - Research suggests the minimum limit can be reduced to 16

### ISSUE THREAD

**Issue by RlzHi** [ankitects/anki#3094]
Hey everyone,

We currently have a minimum of 1000 reviews for FSRS to be optimised, which I believe is now changing to 400. After some research and discussions with @L-M-Sherlock it seems we can lower this as far as 16, or remove it completely.

I've forked and updated srs-benchmark which contains the benchmarking code for this:
https://github.com/RlzHi/srs-benchmark/blob/main/notebook/metric_over_size.ipynb
![download](https://github.com/ankitects/anki/assets/111053265/91eec21c-da44-4bd6-bb69-ab852408e4c2)

The dry run is FSRS 4.5 without optimisations, and after approximately 16 reviews optimising FSRS performs better than the default parameters.

**Comment by L-M-Sherlock** [comment]
@Expertium @user1823, what do you think? I will reproduce the result later.

**Comment by user1823** [comment]
We need to check the stability of the parameters too, i.e., the parameters shouldn't fluctuate too much when re-optimized after doing some more reviews.

**Comment by L-M-Sherlock** [comment]
I also find that `pretrain` could improve the performance significantly when the size of collection is small.

![image](https://github.com/ankitects/anki/assets/32575846/8abd8362-e490-4325-8713-8dbb4b1e7ae0)

user1823's point is interesting. We need to design some experiments to check the stability of parameters.

**Comment by L-M-Sherlock** [comment]
https://github.com/RlzHi/srs-benchmark/blob/3d867c625be926f3f952115e3a957d106c46a6a5/script.py#L191-L201

I notice that the size of testset is only 1/4 of the trainset. It's rare in practice because users wouldn't optimize their parameters frequently.

**Comment by L-M-Sherlock** [comment]
Bad news: I find a severe pitfall in the benchmark code when I design the experiment. I have to re-benchmark all models in this week.

Briefly, the benchmark code has a data leakage problem. The testset is used to pick the best parameters during the training process.

**Comment by Expertium** [comment]
How and why would pretrain work? FSRS already has default parameters built into it.

**Comment by Expertium** [comment]
Regarding the stability of parameters, we could introduce L1 or L2 regularization if n(reviews)<1000 (or some other threshold). It's easy to do with pytorch.

**Comment by Expertium** [comment]
@L-M-Sherlock @user1823 I think there are 2 options:
1) Allow optimization of all parameters even when n(reviews) is very low, say, 100; and introduce L1/L2 regularization when n<1000.
2) If option 1 is difficult to implement, then only do pretrain aka only change the first 4 parameters when 100<=n<=1000. This should be very easy, since the current implementation of pretrain already has a penalty based on the number of reviews.

Btw, I think 16 reviews is too little, I would recommend setting 100 as the new required minimum.
Also, @RlzHi, according to your graph the confidence intervals (I assume that's what the bands are) stop overlapping only around 100 reviews.
![image](https://github.com/ankitects/anki/assets/83031600/90bfd74c-33e6-4d14-9a07-9714e64f8c03)

**Comment by Expertium** [comment]
The specialized tool for calculating accurate confidence intervals is `scipy.stats.bootstrap`, method=BCa.

> Interesting point about stability. Do you think it's important if the parameters change significantly if it achieves a better prediction by doing so?

The problem is that if the sample size is very small (and 16 is small no matter how you look at it, especially since FSRS has 17 parameters), it may lead to overfitting. The algorithm would perform really well on those 16 reviews, and very poorly in practice.

**Comment by Expertium** [comment]
> Optimised FSRS seems to perform better than the default values when n(reviews) ≥ 8

I don't think the graph shows it.

**Comment by dae** [comment]
Personally, I do not think this needs changing. Lower sample sizes are more likely to be noisy, and users who are new to Anki may take a while to be consistent in their grading. Slightly increasing performance for the first few hundred reviews does not seem worth it.

**Comment by Expertium** [comment]
Pretrain is robust even on small sample sizes. Plus, a lot of users have complained about the limit. @L-M-Sherlock what do you think? According to Rlz, pretrain is always better than using default parameters, and the L1 penalty ensures that parameters won't become too big/too small on small sample sizes.

**Comment by dae** [comment]
Thanks for elaborating. The fact that it factors in how the sample size is good. I think you've sold me.

**Comment by Expertium** [comment]
@dae I think it's best if you give us some more time to figure out the details. While it's clear thatp retrain can be helpful for n(reviews)<1000, it's not clear whether there should still be a hard limit - such that below it default parameters are used, no optimization at all - or whether the hard limit should be removed entirely.
So I suggest releasing Anki 24.04 as planned, without changing how the optimizer works compared to the most recent rc. Then, 1 or 2 weeks later, you can release Anki 24.04.1, with the new change.

**Comment by dae** [comment]
Sorry, I meant I was withdrawing my objection, not suggesting we rush this change in.

**Comment by Expertium** [comment]
@dae disregard what I said above, we found a rule that performs striclty better than using default parameters for any sample size.
Pseudocode:
```
old_loss = evaluate(old_parameters, dataset) 

loss = None
parameters = None

loss, parameters = optimize(dataset)  # returns None if optimization failed due to insufficient data
If parameters are None:
    loss, parameters = pretrain_only(dataset)  # returns None if optimization failed due to insufficient data
If parameters are None:
    loss, parameters = evaluate(default_parameters), default_parameters

If old_loss < loss:
    parameters = old_parameters
    loss = old_loss
```

It's unlikely that this will be changed further, so unless you or @L-M-Sherlock have any objections, I recommend implementing this rule in the current version of Anki. Sorry that it's just barely before the deadline, but me and Rlz don't see a good reason to postpone it to the next release.

**Comment by L-M-Sherlock** [comment]
> This rule is complicated, but it guarantees three things:
> 
> 1. Users will never get parameters worse than the default parameters
> 2. Users will never get parameters worse than the previous parameters
> 3. That the best results out of default, pretrain and full optimization are selected

It's unnecessary to evaluate the default parameters. For all new users, the `old_parameters` is `default_parameters`. So the point 2 has ensured the parameters are better than default parameters.

**Comment by Expertium** [comment]
> Which loss?

RMSE.

**Comment by Expertium** [comment]
> It's unnecessary to evaluate the default parameters. For all new users, the old_parameters is default_parameters

But that is true only for the very first optimization.

**Comment by user1823** [comment]
> we found a rule that performs strictly better than using default parameters for any sample size. 

@Expertium, this conclusion is based on the RMSE. But, what about overfitting? Can we be sure that overfitting would not be a significant problem when reviews > 32? Just comparing the RMSE would not help in this case because RMSE would be low when overfitting occurs.

A good way to test overfitting would be to split large datasets based on time and then see how the parameters trained from earlier revlogs perform when applied to later revlogs.

**Comment by Expertium** [comment]
https://github.com/open-spaced-repetition/srs-benchmark/blob/main/notebook/metric_over_size.ipynb

There a lot of graphs. I guess the ones that are the most relevant to your question are these:
![image](https://github.com/ankitects/anki/assets/83031600/a49579cc-075a-4b7b-9b84-9bd1a3851cd2)
Top right: The optimised model is better on train and test
Bottom right: The optimised model is better on train but worse on test
Top left: The optimised model is worse on train but better on test
Bottom left: The optimised model is worse on train and test

**Comment by Expertium** [comment]
Then there is this. 
[![image](https://github.com/ankitects/anki/assets/83031600/07df9f03-12cf-42b6-a3dc-19f8558fc6c9)](https://github.com/open-spaced-repetition/srs-benchmark/blob/main/notebook/minimum_limit.ipynb)

![image](https://github.com/ankitects/anki/assets/83031600/dee86f1d-8e6a-41d1-aa54-5b31f3143838)
Optimized on 8 reviews, tested on the next 10.

![image](https://github.com/ankitects/anki/assets/83031600/b9312ecb-a6a5-431a-a836-27e1773e62f1)
Optimized on 8 reviews, tested on the next 100.

![image](https://github.com/ankitects/anki/assets/83031600/14d2b81c-b0bf-4609-9427-a3ca01125998)
Optimized on 8 reviews, tested on the next 1000.

**Comment by user1823** [comment]
> <img width=400 src="https://private-user-images.githubusercontent.com/83031600/318208484-b9312ecb-a6a5-431a-a836-27e1773e62f1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE4MTUwNzEsIm5iZiI6MTcxMTgxNDc3MSwicGF0aCI6Ii84MzAzMTYwMC8zMTgyMDg0ODQtYjkzMTJlY2ItYTZhNS00MzFhLWE4MzYtMjdlMTc3M2U2MmYxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzMwVDE2MDYxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjMWM4NDMzZGZhOTAwMDk1ODRhNDY1MjVjMzc0YjkzMTY5MWVkYzFiN2Q2YWUxNDY4MzlkODZjYjRkODdiMGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.F9VpWLUccfIT7zH_Z6c64_xOKbQEzCJsZeor_CF1eaA">
>
> Optimized on x reviews, tested on the next 100.

I think that this is the graph we are most interested in. In other words, we can assume that the user would wait for at least 100 reviews before optimizing again. So, we should ensure that the parameters still fit nicely after 100 more reviews.

The confidence bands of dry-run and optimized parameters stop overlapping at about 10<sup>1.8</sup> (≈ 63). So, I think that we should not allow optimization unless there are atleast 63 reviews.

**Comment by Expertium** [comment]
Those are not confidence intervals, but rather 25% and 75% percentiles.

**Comment by user1823** [comment]
> Those are not confidence intervals, but rather 25% and 75% percentiles.

Whatever. Stats is not my field of expertise. But, I hope you got the idea.

**Comment by user1823** [comment]
> I think the takeaway from this is that it doesn't overfit in the short-term, but it does overfit in the long-term if left unoptimised.
> The solution to this seems to be: If the user has optimised on n reviews (e.g. 16), then left the deck unoptimised for 5n reviews (e.g. 100), they should be prompted to re-optimise the deck.

I came to an almost similar conclusion. However, I think that the number should be less than 5n. Maybe 3n or even 2n?

**Comment by Expertium** [comment]
Ok, nevermind, those are, indeed, confidence intervals. My bad.
So it seems that we need around 50-60 reviews for the parameters to work well on the next 100 reviews, and we need around 200 reviews for the parameters ot work well on the next 1000 reviews. So perhaps we should increase the limit for full optimization from 32 to 64 or 128? @RlzHi 
As for how often to reoptimize, I think 2n is good.

**Comment by user1823** [comment]
> we can use pretrain up to 64 reviews

Looking at the graph, even pretrain should not be used before 64 reviews

**Comment by user1823** [comment]
> With 32 reviews, pretrain should perform reasonably up to 132 reviews (100 more).

In the graph with lookahead = 100, the curves and the bands for `pretrain` and `full optimization` appear to be almost overlapping till about 64 reviews. So, I don't think that there should be any difference in the performance of the two.

> If we optimise every time your number of reviews has doubled, the next optimisation will be at 64 reviews (32 more).

I agree. But, then this is valid for full optimization also.

**Comment by Expertium** [comment]
The problem with full optimization is that, _in theory_, it's less stable than pretrain due to a lack of regularization aka penalty for making the parameters too large/too small. Though it doesn't actually show up in data, it seems.

**Comment by user1823** [comment]
It sounds good. I would add that don't even allow pre-train before 32 reviews. The graph with lookahead = 25 shows that about 32 reviews are required to say with 95% certainty that the pretrain parameters would be better than the default ones for the next 25 reviews. This means that below 32 reviews, the user would need to optimize even more frequently than every 2x reviews.

One question:
Does the dataset used to prepare these graphs contain same-day reviews, manual revlogs, etc? If not, Anki would have to use the filtered revlog count for deciding whether to allow optimization or not.

**Comment by Expertium** [comment]
So 32 reviews for pretrain and 64 reviews for the full optimization?

**Comment by user1823** [comment]
> I would rather give the user the choice of optimising in this situation.

Ok. We can remove the limits for pre-train and see how it goes. Because we are talking about very small number of reviews, people will actually be able to share their experience with us during the beta cycle.

**Comment by Expertium** [comment]
> I would rather give the user the choice of optimising in this situation.

I'd rather not. FSRS is already confusing for most people. We should decrease the cognitive burden for the user, not increase it.

**Comment by Expertium** [comment]
Maybe we are misunderstanding each other. I meant that there **shouldn't** be a special window, like "Are you sure you want to run the optimizer when the number of reviews is so low?"

**Comment by Expertium** [comment]
@user1823 here's the answer to your earlier question:
[message (1).txt](https://github.com/ankitects/anki/files/14813470/message.1.txt)
Size: 8
Amt where train loss: fullopt < default < pretrain:    28 / 7586   ~   Test loss: 0.210, 0.193, 0.201
Amt where train loss: fullopt < pretrain < default:   815 / 7586   ~   Test loss: 0.250, 0.251, 0.255
Amt where train loss: default < pretrain < fullopt:     4 / 7586   ~   Test loss: 0.281, 0.220, 0.221
Amt where train loss: default < fullopt < pretrain:    24 / 7586   ~   Test loss: 0.319, 0.317, 0.315
Amt where train loss: pretrain < fullopt < default:    87 / 7586   ~   Test loss: 0.280, 0.281, 0.277
Amt where train loss: pretrain < default < fullopt:     1 / 7586   ~   Test loss: 0.240, 0.228, 0.240

**Comment by user1823** [comment]
@RlzHi, would you be interested in doing another analysis?

Currently, we know that it is good to re-optimize the parameters after 2n reviews. But, is more frequent optimization worth it? For example, if I optimize the parameters on 100k reviews, should I think about re-optimizing them only after I have 200k reviews?

Here is how you can go about it:
- Choose various values of `n`, such as 16, 32, 64, 100, 1000, 10000, 50000, 100000, 200000.
- For each value of `n`, select those collections that have at least `2n` reviews.
- For each value of `n`, optimize the parameters on the first `n` reviews and evaluate them up to the first `2n` reviews.
- For each value of `n`, plot a graph between RMSE and the number of reviews (up to `2n`).
